tests/Doctrine/Tests/DBAL/Functional/StatementTest.php

```diff
@@ -0,0 +1,182 @@
+<?php
+
+namespace Doctrine\Tests\DBAL\Functional;
+
+use Doctrine\DBAL\Schema\Table;
+use Doctrine\DBAL\Types\Type;
+
+class StatementTest extends \Doctrine\Tests\DbalFunctionalTestCase
+{
+    public function testStatementIsReusableAfterClosingCursor()
+    {
+        $sm = $this->_conn->getSchemaManager();
+        $table = new Table('stmt_test_reusable');
+        $table->addColumn('id', 'integer');
+        $sm->createTable($table);
+        $this->_conn->insert('stmt_test_reusable', array('id' => 1));
+        $this->_conn->insert('stmt_test_reusable', array('id' => 2));
+
+        $stmt = $this->_conn->prepare('SELECT id FROM stmt_test_reusable ORDER BY id');
+
+        $stmt->execute();
+
+        $id = $stmt->fetchColumn();
+        $this->assertEquals(1, $id);
+
+        $stmt->closeCursor();
+
+        $stmt->execute();
+        $id = $stmt->fetchColumn();
+        $this->assertEquals(1, $id);
+        $id = $stmt->fetchColumn();
+        $this->assertEquals(2, $id);
+    }
+
+    public function testClosedCursorDoesNotContainResults()
+    {
+        $sm = $this->_conn->getSchemaManager();
+        $table = new Table('stmt_test_no_results');
+        $table->addColumn('id', 'integer');
+        $sm->createTable($table);
+        $this->_conn->insert('stmt_test_no_results', array('id' => 1));
+        $this->_conn->insert('stmt_test_no_results', array('id' => 2));
+
+        $stmt = $this->_conn->prepare('SELECT id FROM stmt_test_no_results');
+
+        $stmt->execute();
+
+        $id = $stmt->fetchColumn();
+        $this->assertNotEmpty($id);
+
+        $stmt->closeCursor();
+
+        try {
+            $value = $stmt->fetchColumn();
+        } catch (\Exception $e) {
+            // some adapters trigger PHP error or throw adapter-specific exception in case of fetching
+            // from a closed cursor, which still proves that it has been closed
```

Comment:
@morozov you are completely right. I feel bad about returning `false` too but unfortunately we have made up that contract when designing the API.

I'm also aware that this PR is growing big already but in the end it is about fixing the close cursor behaviour across all the drivers. Unfortunately it's quite a rabbit hole but it feels wrong to leave some parts buggy when trying to fix a general problem (which this test case exposes). I am ok with a rather big change here if we get it fixed consistently.

Also relying on exceptions from some drivers but not from others in this scenario is not an issue for us because it is undefined behaviour (as not being defined in the interface).